Fix slow back button x-ray flakes

markbastian opened the pull request:
This PR attempts to fix these flaky tests:

- `e2e/test/scenarios/dashboard/dashboard-back-navigation.cy.spec.js > scenarios > dashboard > dashboard back navigation should display a back to the dashboard button in table x-ray dashboards`
- `e2e/test/scenarios/dashboard/dashboard-back-navigation.cy.spec.js > scenarios > dashboard > dashboard back navigation should display a back to the dashboard button in model x-ray dashboards`

Based on [deploysentinel](https://www.deploysentinel.com/ci/analysis?tab=flake), it looks like the problem is the `cy.wait("@dataset");` that occurs after creating a dashboard times our after waiting 5000ms. This PR attempts to fix this issue with the following changes:

- Set the number of generated cards to `const MAX_CARDS = 5;`. This should hopefully substantially reduce the amount of time spent on the FE rendering the dashboard as the dashboard payload should be much smaller.
- Increased the wait timeout on the two offending calls to 15000ms. Hopefully the first solution will always kick in first and this is never needed.

Between these two fixes, I'd hope we'll always see the dataset returned.

markbastian committed:
Attempting to fix x-ray back timeout issues with smaller card count

It looks like the flakes on https://www.deploysentinel.com/ci/analysis?tab=flake for "dashboard back navigation should display a back to the dashboard button in table x-ray dashboards" and "dashboard back navigation should display a back to the dashboard button in model x-ray dashboards" frequently fail at the `cy.wait("@dataset")` stage. I'm guessing this is because the dashboard code is so slow. This PR limits the number of cards produce to `MAX_CARDS` (5) in order to speed up card generation and help with the flakes.

markbastian committed:
Adding a timeout to slow x-rays for back button. Hopefully, the small number of max cards makes this unnecessary, though.

markbastian committed:
Bumped timeout to 15s instead of 5s. Looks like it was already 5. Again, hopefully the change to minimized the number of generated cards will fix this flake and not take as long.

iethree commented on e2e/test/scenarios/dashboard/dashboard-back-navigation.cy.spec.js:
```diff
@@ -104,8 +106,8 @@ describe("scenarios > dashboard > dashboard back navigation", () => {
       const cardTitle = "Sales per state";
-      cy.visit(`/auto/dashboard/table/${ORDERS_ID}`);
-      cy.wait("@dataset");
+      cy.visit(`/auto/dashboard/table/${ORDERS_ID}?show=${MAX_CARDS}`);
```
I can't get this `show` query param to do anything 🤔

markbastian replied:
Did some digging into the API vs. code code. Core uses `show` to limit cards, but the API doesn't actually pass it through, so this isn't a good fix ATM. WDYT of upping the timeouts on those two calls? That _seems_ to be the issue.

markbastian replied:
Pushed a change and now it should work.

iethree replied:
Upping the timeouts is probably a fine band-aid to reduce the flakes but I've seen some of these API requests hit a minute or more, so it doesn't really address the issue.

markbastian replied:
I suspect the timing is just because of the amount of time required to hit every endpoint for every dashcard. I think only showing a few cards should fix this. From a product perspective, perhaps someone only wants to see the top 2-3 cards.

markbastian committed:
Fixing the `show` query param for x-rays

`show` in `core` can be either `:all` or an integer. However, the API layer only exposes the `:all` or `nil` options. This PR exposes the integer path as well.

Note that this is incremental to get flakes working faster. I'll still need to add API tests for this.

markbastian commented:
Fixed `show` query param behavior.
![show](https://github.com/metabase/metabase/assets/8381441/40de6eef-f241-43c7-8f47-4a406856ec96)

iethree commented:
I hate to expose APIs just for the sake of testing 😬  Is there a product reason to expose this `show` query Parameter? What do you think @luizarakaki ?

markbastian committed:
Adding API unit tests to `metabase.api.automagic-dashboards-test` for `show` cases.

markbastian commented:
> I hate to expose APIs just for the sake of testing 😬 Is there a product reason to expose this `show` query Parameter? What do you think @luizarakaki ?

I pushed tests up for this, so we should be good to go if we either:
- Decide to expose `show` n cards. Note that the underlying domain api already supports and tests the numeric limit count. I am only making it visible at the API.
- Do not expose `show` n cards. If we do, I can just revert all changes except for the timeout increase (as long as @iethree and co. think it's a reasonable test fix).

markbastian committed:
Fixed deftest name to be unique.

markbastian committed:
Fixing reference to new card in flaky test
